raft/raft.go

```diff
@@ -423,10 +475,59 @@ func (r *raft) send(m pb.Message) {
 		}
 	}
-	if m.To == r.id {
-		r.logger.Panicf("message should not be self-addressed when sending %s", m.Type)
+	if m.Type == pb.MsgAppResp || m.Type == pb.MsgVoteResp || m.Type == pb.MsgPreVoteResp {
+		// If async storage writes are enabled, messages added to the msgs slice
+		// are allowed to be sent out before unstable state (e.g. log entry
+		// writes and election votes) have been durably synced to the local
+		// disk.
+		//
+		// For most message types, this is not an issue. However, response
+		// messages that relate to "voting" on either leader election or log
+		// appends require durability before they can be sent. It would be
+		// incorrect to publish a vote in an election before that vote has been
+		// synced to stable storage locally. Similarly, it would be incorrect to
+		// acknowledge a log append to the leader before that entry has been
+		// synced to stable storage locally.
+		//
+		// Per the Raft thesis, section 3.8 Persisted state and server restarts:
+		//
+		// > Raft servers must persist enough information to stable storage to
+		// > survive server restarts safely. In particular, each server persists
+		// > its current term and vote; this is necessary to prevent the server
+		// > from voting twice in the same term or replacing log entries from a
+		// > newer leader with those from a deposed leader. Each server also
+		// > persists new log entries before they are counted towards the entries’
+		// > commitment; this prevents committed entries from being lost or
+		// > “uncommitted” when servers restart
+		//
+		// To enforce this durability requirement, these response messages are
+		// queued to be sent out as soon as the current collection of unstable
+		// state (the state that the response message was predicated upon) has
+		// been durably persisted. This unstable state may have already been
+		// passed to a Ready struct whose persistence is in progress or may be
+		// waiting for the next Ready struct to begin being written to Storage.
+		// These messages must wait for all of this state to be durable before
+		// being published.
+		//
+		// Rejected responses (m.Reject == true) present an interesting case
+		// where the durability requirement is less unambiguous. A rejection may
+		// be predicated upon unstable state. For instance, a node may reject a
+		// vote for one peer because it has already begun syncing its vote for
+		// another peer. Or it may reject a vote from one peer because it has
+		// unstable log entries that indicate that the peer is behind on its
+		// log. In these cases, it is likely safe to send out the rejection
+		// response immediately without compromising safety in the presence of a
+		// server restart. However, because these rejections are rare and
+		// because the safety of such behavior has not been formally verified,
+		// we err on the side of safety and omit a `&& !m.Reject` condition
+		// above.
+		r.msgsAfterAppend = append(r.msgsAfterAppend, m)
+	} else {
```

Comment:
This is a reordering of messages relative to the status quo: messages other than voting responses now get to cut in line and go ahead of (earlier) voting responses. That should be fine for correctness (raft is already theoretically robust against reordered messages), but preserving order of messages has proven important for performance in the past. 

This was mainly about ordering within a message type, although I think there may have been interactions between appends and snapshots (notable since only appends are affected by this change). I can't think of any problems here but just wanted to raise it as a possible concern.

Reply:
This is a good point. As you mention, we now introduce certain re-ordering across message types, but not within a message type. I haven't been able to construct a case where delaying any of the three `msgsAfterAppend` message types causes issues.

>  I think there may have been interactions between appends and snapshots

Was this between MsgApp and MsgSnap, or MsgAppResp and MsgSnap? It must have been the former, as MsgApp and MsgSnap are both sent from the leader, while MsgAppResp is sent from followers.

Also, recall that MsgSnap also uses a MsgAppResp to respond to the leader, so these responses will remain ordered.